### PR TITLE
Add VTK/9.0.1

### PR DESCRIPTION
**Pull request by rockandsalt**
Specify library name and version:  **vtk/9.0.1**

I am creating a pull request to start and track work on adding the library VTK to Conan Center. VTK is a common visualization library used in many industries. It would be nice to have it in conan center. Current pull request is not ready for merge.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by rockandsalt** [comment]
I just noticed a [stale pull request for vtk](stale pull request for vtk). I will try to work from it.

**Comment by rockandsalt** [comment]
@prince-chrismc, I saw you reviewed a previous pull request. VTK is a huge library composed of a lot of modules. But it has the particularity where each module comes with a vtk.module file ([example](https://github.com/Kitware/VTK/blob/master/Common/DataModel/vtk.module)) which list the name and the dependencies of this module.  I am thinking of copying those files and model the library through the [components.](https://docs.conan.io/en/1.36/creating_packages/package_information.html#package-information-components). I think it would be cleaner than listing all the modules manually. 

But, I am not sure how stable it is and how the namespace of it works. The top library is named `VTK`, is each component named `VTK::Component` or does conan append the `VTK::` automatically?

**Comment by prince-chrismc** [comment]
You caught  me at a bad week since I am fairly busy.

The best recommendation I can make is following the approach for the Qt project. We have a file that describes all the components and we use that to generate them.

We have full control of the CMake targets with or without namespace.

Here an example of a complicated cmake export https://github.com/conan-io/conan-center-index/blob/f8a156cc061afe6ec7c319619c897c7ab95d4d17/recipes/libgeotiff/all/conanfile.py

File: recipes/vtk/all/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.12...3.16 FATAL_ERROR)
```
Comment (prince-chrismc) [inline recipes/vtk/all/CMakeLists.txt:1]:
```suggestion
cmake_minimum_required(VERSION 3.4)
```

File: recipes/vtk/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    name = "vtk"
+    license = "BSD-3-Clause"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://vtk.org/"
+    description = "VTK is an open-source software system for" \
+        "image processing, 3D graphics, volume rendering and visualization."
+    topics = ("conan", "vtk", "geometry", "algorithms")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False],
+               "build_testing": ["ON", "OFF", "WANT"],
```
Comment (prince-chrismc) [inline recipes/vtk/all/conanfile.py:16]: This should always be off in CCI

File: recipes/vtk/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+               "vtk_smp_implementation": ["Sequential", "OpenMP", "TBB", "STDThread"],
+               "vtk_wrap_python": [True, False],
+               "vtk_use_cuda": [True, False],
+               "vtk_use_mpi": [True, False],
+               "vtk_group_enable_qt": ["YES", "WANT", "DONT_WANT", "NO", "DEFAULT"],
```
Comment (prince-chrismc) [inline recipes/vtk/all/conanfile.py:21]: What is the difference between yes/want and no/dont_want? and default should be the default option?

File: recipes/vtk/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                       "vtk_group_enable_imaging": "DEFAULT",
+                       "vtk_group_enable_qt": "DEFAULT",
+                       "vtk_group_enable_rendering": "WANT",
+                       "vtk_group_enable_mpi": "DONT_WANT",
+                       "vtk_group_enable_standalone": "WANT",
+                       "vtk_group_enable_views": "DEFAULT",
+                       "vtk_group_enable_web": "DEFAULT",
+                       "fPIC": True}
+    generators = "cmake"
+    version = "9.0.1"
```
Comment (prince-chrismc) [inline recipes/vtk/all/conanfile.py:44]: you can pass this at the command line

```suggestion
```

File: recipes/vtk/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            del self.options.fPIC
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def source(self):
+        if self.version is None:
+            self.version = "9.0.1"
+        tools.get(**self.conan_data["sources"][self.version],
```
Comment (prince-chrismc) [inline recipes/vtk/all/conanfile.py:63]:
```suggestion
        tools.get(**self.conan_data["sources"][self.version],
```

File: recipes/vtk/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                  strip_root=True, destination=self._source_subfolder)
+
+    def _configure_cmake(self):
+        if not self._cmake:
+            self._cmake = CMake(self)
+            self._cmake.definitions["BUILD_SHARED_LIBS"] = self.options.shared
+            self._cmake.definitions["BUILD_TESTING"] = self.options.build_testing
```
Comment (prince-chrismc) [inline recipes/vtk/all/conanfile.py:70]:
```suggestion
            self._cmake.definitions["BUILD_TESTING"] = False
```

File: recipes/vtk/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+               "vtk_group_enable_imaging": ["YES", "WANT", "DONT_WANT", "NO", "DEFAULT"],
+               "fPIC": [True, False]}
+    default_options = {"shared": False,
+                       "build_testing": "OFF",
+                       "vtk_smp_implementation": "Sequential",
+                       "vtk_wrap_python": False,
+                       "vtk_use_cuda": False,
+                       "vtk_use_mpi": False,
+                       "vtk_group_enable_imaging": "DEFAULT",
+                       "vtk_group_enable_qt": "DEFAULT",
```
Comment (prince-chrismc) [inline recipes/vtk/all/conanfile.py:36]: Our convention is `with_<dep>` 

https://github.com/conan-io/conan-center-index/blob/master/docs/reviewing.md#recommended-feature-options-names

File: recipes/vtk/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+            self._cmake.definitions["VTK_WRAP_PYTHON"] = self.options.vtk_wrap_python
+            self._cmake.definitions["VTK_USE_CUDA"] = self.options.vtk_use_cuda
+            self._cmake.definitions["VTK_GROUP_ENABLE_Imaging"] = self.options.vtk_group_enable_imaging
+            self._cmake.definitions["VTK_GROUP_ENABLE_Qt"] = self.options.vtk_group_enable_qt
+            self._cmake.definitions["VTK_GROUP_ENABLE_Rendering"] = self.options.vtk_group_enable_rendering
+            self._cmake.definitions["VTK_GROUP_ENABLE_MPI"] = self.options.vtk_group_enable_mpi
+            self._cmake.definitions["VTK_GROUP_ENABLE_StandAlone"] = self.options.vtk_group_enable_standalone
+            self._cmake.definitions["VTK_GROUP_ENABLE_Views"] = self.options.vtk_group_enable_views
+            self._cmake.definitions["VTK_GROUP_ENABLE_Web"] = self.options.vtk_group_enable_web
+            self._cmake.configure(source_folder=self._source_subfolder)
```
Comment (prince-chrismc) [inline recipes/vtk/all/conanfile.py:81]: This skips the CMake wrapper at the top which might explain why it failed in some configuration

File: recipes/vtk/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    license = "BSD-3-Clause"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://vtk.org/"
+    description = "VTK is an open-source software system for" \
+        "image processing, 3D graphics, volume rendering and visualization."
+    topics = ("conan", "vtk", "geometry", "algorithms")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False],
+               "build_testing": ["ON", "OFF", "WANT"],
+               "vtk_smp_implementation": ["Sequential", "OpenMP", "TBB", "STDThread"],
```
Comment (prince-chrismc) [inline recipes/vtk/all/conanfile.py:17]: Option values should be lower case https://github.com/conan-io/conan-center-index/blob/41d4fbb8c218f9d938ccb8b9d888bff52002e62e/recipes/muparser/all/conanfile.py#L18

```suggestion
               "vtk_smp_implementation": [None, "openmp", "tbb", "stl"],
```

https://github.com/conan-io/conan-center-index/blob/41d4fbb8c218f9d938ccb8b9d888bff52002e62e/recipes/libgettext/all/conanfile.py#L18

There's no good choice for `std::thread`

File: recipes/vtk/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+    def requirements(self):
+        if self.options.vtk_group_enable_rendering in ["YES", "WANT", "DEFAULT"]:
+            self.requires("opengl/system")
+
+        if self.options.vtk_group_enable_qt in ["YES", "WANT", "DEFAULT"]:
+            self.requires("qt/5.15.2")
+
+        if self.options.vtk_smp_implementation == "TBB":
+            self.requires("TBB/[>2020]")
```
Comment (prince-chrismc) [inline recipes/vtk/all/conanfile.py:97]: we do not use version range in CCI

File: recipes/vtk/all/test_package/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+find_package(VTK REQUIRED)
+
+add_executable(example example.cpp)
+target_link_libraries(example ${CONAN_LIBS} ${VTK_LIBRARIES})
+
+# CTest is a testing tool that can be used to test your project.
+# enable_testing()
+# add_test(NAME example
+#          WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/bin
+#          COMMAND example)
```
Comment (prince-chrismc) [inline recipes/vtk/all/test_package/CMakeLists.txt:15]:
```suggestion
target_link_libraries(example ${CONAN_LIBS} ${VTK_LIBRARIES})
```

File: recipes/vtk/all/test_package/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if not tools.cross_building(self):
+            os.chdir("bin")
+            self.run(".%sexample" % os.sep)
```
Comment (prince-chrismc) [inline recipes/vtk/all/test_package/conanfile.py:18]:
```suggestion
            self.run(os.path.join("bin", "example"), run_environment=True)
```

File: recipes/vtk/all/conandata.yml [hunk 13]
```diff
@@ -0,0 +1,4 @@
+sources:
+  "9.0.1":
+    sha256: a917141eed21d13cb51db383ae9a1947ed885aedf37f45d069d42b6698ac3666
+    url: https://github.com/Kitware/VTK/archive/refs/tags/v9.0.1.zip
```
Comment (prince-chrismc) [inline recipes/vtk/all/conandata.yml:4]: we prefer a `tar.gz` if it's available

**Review by prince-chrismc — COMMENTED** [review]
Sorry about the delay, lots of comments but overall it's good

**Comment by rockandsalt** [comment]
> Sorry about the delay, lots of comments but overall it's good

No worries, work got busy as well.

File: recipes/vtk/all/conanfile.py (hunk 3 above)
Comment (rockandsalt) [reply recipes/vtk/all/conanfile.py:21]: Yeah, I am not too clear about that as well. This just mirror the vtk cmake options. I think it has to do with priority level.

File: recipes/vtk/all/conanfile.py (hunk 2 above)
Comment (rockandsalt) [reply recipes/vtk/all/conanfile.py:16]: you mean build test?

Comment (prince-chrismc) [reply recipes/vtk/all/conanfile.py:16]: Yes! This option should be removed

File: recipes/vtk/all/conanfile.py (hunk 3 above)
Comment (prince-chrismc) [reply recipes/vtk/all/conanfile.py:21]: Maybe just True/False keep it simple? Use something like https://github.com/conan-io/conan-center-index/blob/41d4fbb8c218f9d938ccb8b9d888bff52002e62e/recipes/wolfssl/all/conanfile.py#L99
to pass them if it's required

File: recipes/vtk/all/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+            dep = self.filter_dep(name, content['depends'], enabled_pkg)
+            opt_dep = self.filter_dep(name, content['private_depends'], enabled_pkg)
+            self.cpp_info.components[alias].requires.extend(dep)
+            self.cpp_info.components[alias].requires.extend(opt_dep)
+            
+            if name == "VTK::GUISupportQt" :
+                self.cpp_info.components[alias].requires.append("qt")
+            
+            if name == "VTK::RenderingOpenGL2":
+                self.cpp_info.components[alias].requires.append("opengl/system")
```
Comment (rockandsalt) [inline recipes/vtk/all/conanfile.py:174]: @prince-chrismc , I found some time this week to work on this. I feel I am getting close. But the CI throws an error about not finding opengl. I added this line and opengl does get added to the list of requires. But conan still throws and error about ` package require 'opengl' not used in components requires `. Did I do it right?

Comment (rockandsalt) [inline recipes/vtk/all/conanfile.py:174]: @prince-chrismc , I found some time this week to work on this. I feel I am getting close. But the CI throws an error about not finding opengl. I added this line and opengl does get added to the list of requires. But conan still throws and error about package require 'opengl' not used in components requires. Did I do it right?

File: recipes/vtk/all/conanfile.py [hunk 15]
```diff
@@ -58,26 +55,45 @@ def config_options(self):
...
+        if v == "sequential":
+            return "Sequential"
+        elif v == "openmp":
+            return "OpenMP"
+        elif v == "tbb":
+            return "TBB"
+        elif v == "stl":
+            return "STDThread"
+        else:
+            raise "Invalid SMP Option"
```
Comment (prince-chrismc) [inline recipes/vtk/all/conanfile.py:71]: I recall this being better for the cli to handle

```suggestion
            raise ConanException("Invalid SMP Option")
```

File: recipes/vtk/all/conanfile.py [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+                       "vtk_use_mpi": False,
+                       "vtk_group_enable_imaging": True,
+                       "vtk_group_enable_qt": True,
+                       "vtk_group_enable_rendering": True,
+                       "vtk_group_enable_mpi": False,
+                       "vtk_group_enable_standalone": True,
+                       "vtk_group_enable_views": "default",
+                       "vtk_group_enable_web": False,
+                       "fPIC": True}
+    generators = "cmake"
```
Comment (prince-chrismc) [inline recipes/vtk/all/conanfile.py:40]:
```suggestion
    generators = "cmake", "cmake_find_package_multi"
```

**Review by prince-chrismc — COMMENTED** [review]
Looking into the error build error, this project has a lot of custom behavior (kitware is very good at CMake)... in https://github.com/Kitware/VTK/tree/master/CMake I would suggest deleting the `Find*.cmake` since they can collide with Conans (one you enable the generator)

***

I have a personal family matter that I am dealing with so I would encourage you to ping one of the other reviewers for help, they are listed here https://github.com/conan-io/conan-center-index/blob/master/docs/review_process.md#getting-your-pull-request-reviewed

This looks like a tough one and I am not available to help... but best of luck!

**Comment by rockandsalt** [comment]
@prince-chrismc no worries thank you for the tip.

**Pull request closed without merging**